cuda/base/kernel_launch.cuh

```diff
@@ -0,0 +1,256 @@
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CUDA_BASE_KERNEL_LAUNCH_CUH_
+#define GKO_CUDA_BASE_KERNEL_LAUNCH_CUH_
+
+
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+#include "cuda/base/device_guard.hpp"
+#include "cuda/base/types.hpp"
+#include "cuda/components/thread_ids.cuh"
+
+
+#ifdef GKO_KERNEL
+#error "Only one kernel_launch.hpp file can be included at a time."
+#else
+#define GKO_KERNEL __device__
+#endif
+
+
+namespace gko {
+namespace kernels {
+namespace cuda {
+
+
+constexpr auto default_block_size = 512;
+
+
+template <typename ValueType>
+struct matrix_accessor {
+    ValueType *data;
+    size_type stride;
+
+    __device__ ValueType &operator()(size_type row, size_type col)
+    {
+        return data[row * stride + col];
+    }
+
+    __device__ ValueType &operator[](size_type idx) { return data[idx]; }
+};
+
+
+template <typename T>
+struct to_device_type_impl {
+    using type = std::decay_t<cuda_type<T>>;
+    static type map_to_device(T in) { return as_cuda_type(in); }
+};
+
+template <typename ValueType>
+struct to_device_type_impl<matrix::Dense<ValueType> *&> {
+    using type = matrix_accessor<cuda_type<ValueType>>;
+    static type map_to_device(matrix::Dense<ValueType> *mtx)
+    {
+        return {as_cuda_type(mtx->get_values()), mtx->get_stride()};
+    }
+};
+
+template <typename ValueType>
+struct to_device_type_impl<const matrix::Dense<ValueType> *&> {
+    using type = matrix_accessor<const cuda_type<ValueType>>;
```

Comment:
Yes, definitely. The main reason I had to build a separate accessor is that the existing Ginkgo ranges have value semantics, i.e. they would copy the data by default, instead of just copying pointers. That would make writing kernels much more fragile.